File: runtime/observatory/lib/src/models/objects/thread.dart
// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

part of models;

enum ThreadKind {
  unknownTask,
  mutatorTask,
  compilerTask,
  sweeperTask,
  markerTask,
  finalizerTask
}

abstract class Thread {
  /// The id associated with the thread on creation.
  String get id;

  /// The task type associated with the thread.
  ThreadKind get kind;

  /// The maximum amount of memory in bytes allocated by a thread at a given
  /// time throughout the entire life of the thread.
  int get memoryHighWatermark;

  /// A list of all the zones held by the thread.
  Iterable<Zone> get zones;
}

File: runtime/bin/thread.cc
// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include "bin/thread.h"

#include "platform/utils.h"

namespace dart {
namespace bin {

void Thread::Start(const char* name,
                   ThreadStartFunction function,
                   uword parameter) {
  int result = TryStart(name, function, parameter);
  if (result != 0) {
    const int kBufferSize = 1024;
    char error_buf[kBufferSize];
    FATAL("Could not start thread %s: %d (%s)", name, result,
          Utils::StrError(result, error_buf, kBufferSize));
  }
}

}  // namespace bin
}  // namespace dart
